test/unit/kubernetes-deploy/kubernetes_resource/deployment_test.rb

```diff
@@ -2,118 +2,277 @@
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
-      }
+  def test_deploy_succeeded_with_max_unavailable_as_percent
+    status = {
+      "replicas" => 3,
+      "updatedReplicas" => 2,
+      "unavailableReplicas" => 2,
+      "availableReplicas" => 1
+    }
+
+    new_rs_status = {
+      "replicas" => 2,
+      "availableReplicas" => 1,
+      "readyReplicas" => 1
     }
+    deploy = build_synced_deployment(status: status, new_rs_status: new_rs_status,
+      rollout: 'maxUnavailable', max_unavail: "100%")
+    assert deploy.deploy_succeeded?
+
+    deploy = build_synced_deployment(status: status, new_rs_status: new_rs_status,
+      rollout: 'maxUnavailable', max_unavail: "67%") # rounds up to two max
+    assert deploy.deploy_succeeded?
 
-    deploy = KubernetesDeploy::Deployment.new(namespace: "", context: "", logger: logger, definition: rollout)
-    mock_rs = Minitest::Mock.new
-    needed = 2
-    mock_rs.expect :present?, true
-    mock_rs.expect :desired_replicas, needed
-    mock_rs.expect :ready_replicas, needed - 1
-    mock_rs.expect :available_replicas, needed - 1
-    deploy.instance_variable_set(:@max_unavailable, 0)
-    deploy.instance_variable_set(:@latest_rs, mock_rs)
-    deploy.instance_variable_set(:@desired_replicas, needed)
+    deploy = build_synced_deployment(status: status, new_rs_status: new_rs_status,
+      rollout: 'maxUnavailable', max_unavail: "66%") # rounds down to one max
+    refute deploy.deploy_succeeded?
 
+    deploy = build_synced_deployment(status: status, new_rs_status: new_rs_status,
+      rollout: 'maxUnavailable', max_unavail: "0%")
     refute deploy.deploy_succeeded?
   end
 
-  def test_deploy_succeeded_fails_with_max_unavailable_as_a_percent
-    rollout = {
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
-      }
-    }
+  def test_deploy_succeeded_raises_with_invalid_rollout_annotation
+    deploy = build_synced_deployment(status: {}, new_rs_status: {}, rollout: 'bad', max_unavail: "33%")
+    m = "'kubernetes-deploy.shopify.io/required-rollout: bad' is invalid. Acceptable values: maxUnavailable, full, none"
+    assert_raises_message(KubernetesDeploy::FatalDeploymentError, m) do
+      deploy.deploy_succeeded?
+    end
+  end
+
+  def test_validation_fails_with_invalid_rollout_annotation
+    deploy = build_synced_deployment(status: {}, new_rs_status: {}, rollout: 'bad', max_unavail: false)
+    deploy.kubectl.expects(:run).with('create', '-f', anything, '--dry-run', '--output=name', anything).returns(
+      ["", "super failed", SystemExit.new(1)]
+    )
+    refute deploy.validate_definition
+
+    expected = <<~STRING.strip
+      super failed
+      'kubernetes-deploy.shopify.io/required-rollout: bad' is invalid. Acceptable values: maxUnavailable, full, none
```

Comment:
Its a bit brittle to test against a string. You could expect `rollout_annotation_err_msg` to be called, though I have mixed feelings about testing for a private function being called too

Reply:
True, but IMO the value of asserting something about it is seeing that the full string makes sense. I actually caught a few errors when I wrote these assertions (spacing/punctuation problems, but more importantly that "is invalid with strategy 'rollingUpdate'" was saying the opposite of what it meant, which I didn't realize until I read the whole thing in the test). I can at least reference the constant instead of hard-coding the annotation though.